### PR TITLE
handle case where all electrodes have same coordinate, eg (0, 0)

### LINKED ISSUE #292: handle case where all electrodes have same coordinate, eg (0, 0)

### PR DESCRIPTION
**Pull request by magland**
Fixes #292

@jsoules. This started out as a simple fix to the degenerate case of all electrodes having the same coordinates. But then I found that ElectrodeGeometryWidget did not use the same code as the other widgets. So then it took some work to get that to use the same layers. That's why there are little adjustments to the types throughout. Nothing major.

### CODE REVIEW

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts [hunk 1]
```diff
@@ -102,7 +102,22 @@ const setupVerticalElectrodes = ({width, height, electrodeIds}: {width: number,
...
 
-const setupElectrodes = (args: {width: number, height: number, electrodeLocations: Vec2[], electrodeIds: number[], layoutMode: 'geom' | 'vertical'}): {
+const fixDegenerateCase = (electrodeLocations: Vec2[]) => {
+    const box = {
+        xmin: getArrayMin(electrodeLocations.map(e => (e[0]))),
+        xmax: getArrayMax(electrodeLocations.map(e => (e[0]))),
+        ymin: getArrayMin(electrodeLocations.map(e => (e[1]))),
+        ymax: getArrayMax(electrodeLocations.map(e => (e[1])))
+    }
+    if ((box.xmin === box.xmax) && (box.ymin === box.ymax)) {
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:112]: So this still works okay if one of the dimensions is zero? (I see it's mapped to a linear geometry, but I was also wondering what would happen if the electrodes were actually laid out in a line anyway)

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts [hunk 2]
```diff
@@ -112,20 +127,21 @@ const setupElectrodes = (args: {width: number, height: number, electrodeLocation
...
+    const radius = computeRadius(electrodeLocations2)
+    let boundingBox = getElectrodesBoundingBox(electrodeLocations2, radius)
     let boxAspect = getWidth(boundingBox) / getHeight(boundingBox)
 
-    let realizedElectrodeLocations = electrodeLocations
+    let realizedElectrodeLocations = electrodeLocations2
     if ((boxAspect > 1) !== (canvasAspect > 1)) {
         // if the two aspect ratios' relationship to 1 is different, then one is portrait
         // and the other landscape. We should then correct by rotating the electrode set 90 degrees.
         // note: a 90-degree right rotation in 2d makes x' = y and y' = -x
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:143]: A 90-degree right rotation *about the origin*, I meant here.

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts [hunk 3]
```diff
@@ -112,20 +127,21 @@ const setupElectrodes = (args: {width: number, height: number, electrodeLocation
     if (layoutMode === 'vertical') {
         return setupVerticalElectrodes({width, height, electrodeIds})
     }
+    const electrodeLocations2 = fixDegenerateCase(electrodeLocations)
```
Comment (jsoules) [inline src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:130]: Might be more descriptive to call this something like 'correctedElectrodeLocations', or just have it be `let realizedElectrodeLocations` (l. 139) from the start

**Review by jsoules — COMMENTED** [review]
As per your note, a lot of these changes are standardizing the way that electrode geometry is handled to comply with the other widgets; this is definitely a good thing.

I had a couple notes on the specific fix, but nothing urgent.

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts (hunk 1 above)
Comment (magland) [reply src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:112]: I don't think it's a problem if they are laid out linearly. The estimated spacing is then non-zero.

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts (hunk 2 above)
Comment (magland) [reply src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:143]: i'll update that

File: src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts (hunk 3 above)
Comment (magland) [reply src/extensions/averagewaveforms/AverageWaveformsView/setupElectrodes.ts:130]: you're right, I'll update that

**Pull request merged**